### PR TITLE
Building: Build Unix static libraries a limited number of object files at a time

### LINKED ISSUE #12116: MinGW build fails due to overlong ar argument list
My recent test build using MSYS2/MinGW msys2-x86_64-20200602 failed like given below.
Apparently the number of object files exceeded a limit of that `ar` instance..
````
openssl/master-3.0.0-alpha3
$ make
make depend && make _build_sw
make[1]: Entering directory '/p/PKI/openssl/master-3.0.0-alpha3'
make[1]: Leaving directory '/p/PKI/openssl/master-3.0.0-alpha3'
make[1]: Entering directory '/p/PKI/openssl/master-3.0.0-alpha3'
gcc  -I. -Iinclude -Iapps/include  -m64 -Wall -O3 -DL_ENDIAN -DOPENSSL_PIC -DOPENSSLDIR="\"/usr/local/ssl\"" -DENGINESDIR="\"/usr/local/lib/engines-3\"" -DMODULESDIR="\"/usr/local/lib/ossl-modules\"" -DUNICODE -D_UNICODE -DWIN32_LEAN_AND_MEAN -D_MT -DOPENSSL_BUILDING_OPENSSL -DNDEBUG  -MMD -MF apps/lib/libapps-lib-app_params.d.tmp -MT apps/lib/libapps-lib-app_params.obj -c -o apps/lib/libapps-lib-app_params.obj apps/lib/app_params.c

(many more lines)

gcc  -Iproviders/common/include/prov -Icrypto -Iinclude -Iproviders/implementations/include -Iproviders/common/include  -m64 -Wall -O3 -DL_ENDIAN -DOPENSSL_PIC -DOPENSSLDIR="\"/usr/local/ssl\"" -DENGINESDIR="\"/usr/local/lib/engines-3\"" -DMODULESDIR="\"/usr/local/lib/ossl-modules\"" -DUNICODE -D_UNICODE -DWIN32_LEAN_AND_MEAN -D_MT -DOPENSSL_BUILDING_OPENSSL -DNDEBUG  -MMD -MF providers/implementations/signature/libnonfips-lib-rsa.d.tmp -MT providers/implementations/signature/libnonfips-lib-rsa.obj -c -o providers/implementations/signature/libnonfips-lib-rsa.obj providers/implementations/signature/rsa.c
ar r libcrypto.a crypto/aes/libcrypto-lib-aes-x86_64.obj crypto/aes/libcrypto-lib-aes_cfb.obj crypto/aes/libcrypto-lib-aes_ecb.obj crypto/aes/libcrypto-lib-aes_ige.obj crypto/aes/libcrypto-lib-aes_misc.obj 

(many more) 

crypto/aes/libcrypto-lib-aes_ofb.obj crypto/aes/libcrypto-lib-aes_wrap.obj providers/implementations/signature/libnonfips-lib-rsa.obj
make[1]: ar: Argument list too long
make[1]: *** [Makefile:12469: libcrypto.a] Error 127
make[1]: Leaving directory '/p/PKI/openssl/master-3.0.0-alpha3'
make: *** [Makefile:3265: build_sw] Error 2
```

### PR DESCRIPTION
**Pull request by levitte**
We're hitting problems that the 'ar' command line becomes too long for
some 'make' versions, or the shell it uses.

We therefore change the way we create a static library by doing so one
object file at a time.  This is slower, but has better guarantees to
work properly on limited systems.

Fixes #12116

### CODE REVIEW

**Comment by petrovr** [comment]
Richard Levitte wrote:
> We&#39;re hitting problems that the &#39;ar&#39; command line becomes too long for
> some &#39;make&#39; versions, or the shell it uses.
>
> We therefore change the way we create a static library by doing so one
> object file at a time.  This is slower, but has better guarantees to
> work properly on limited systems.
No response files?

Functionality exist event before gcc .

Roumen

**Comment by kaduk** [comment]
> This is slower

Isn't it quadratically slower in some cases?

> The addition of 1.1.1 is tentative. Do we want that?

I would be inclined to keep 1.1.1 and 3.0 in sync, whatever solution we end up with.

**Comment by levitte** [comment]
@petrovr, unless we know that response file is a feature in every `ar` where openssl gets built, then it becomes difficult.
I do have some ideas about defining toolchain capabilities with better granularity than what we have today, though, and that might help... It will take a bit of time to get that together, though...

**Comment by richsalz** [comment]
Ugh.  How about at least one command per subdirectory?

**Comment by richsalz** [comment]
Also, you have the complete list of object files in the Perl stuff, make multiple ar commands of 50-100 per invocation.

**Comment by rsbeckerca** [comment]
Didn't the fix for the NonStop port address (Issue #5087 ) this situation?

**Comment by levitte** [comment]
@richsalz, one commend per subdirectory could be done, sure, and would probably be enough to shorten command lines enough.

@rsbeckerca, did the non-stop work touch this particular bit? My pardon, I've had much focus elsewhere...

**Comment by rsbeckerca** [comment]
@levitte We had the same issue, yes. Our 1.1.1g is building/testing well at this point.

**Comment by DDvO** [comment]
> Note: The addition of 1.1.1 is tentative. Do we want that?

It's likely not needed there because so far (before starting with 3.0.0) the number of object files apparently was below the limit
since so far this issue has not been reported.

**Comment by DDvO** [comment]
> @DDvO, please try this

I'm running it now, but that installation it terribly slow, will let you know when done...

**Comment by DDvO** [comment]
> > This is slower
> 
> Isn't it quadratically slower in some cases?

Likely not - the name of the `q` option used now is derived from "quick" and its man page says:

> Quick append; Historically, add the files member... to the end of archive, without checking for replacement.
> 
> [...]        
>            Since the point of this operation is speed, implementations of ar have the option of not updating the archive's symbol table if one exists.  Too many
>            different systems however assume that symbol tables are always up-to-date, so GNU ar will rebuild the table even with a quick append.
>

**Comment by DDvO** [comment]
> > @DDvO, please try this
> 
> I'm running it now, but that installation it terribly slow, will let you know when done...

The solution works but can be pretty inefficient.

Due to the very bad overall file system performance of the Windoze system where my MSYS2 is installed,
every part of the build is very slow. The addition of each object file takes on average some 2-3 seconds,
and this deteriorates to about 5 seconds towards the end of filling the archive.
In total this adds more than 30 minutes to the build time, which I do not think is bearable.

**Comment by DDvO** [comment]
> Ugh. How about at least one command per subdirectory?

It may be easier to split the list into...
* a fixed number of rather equally long chunks (just two halves should be sufficient) or 
* sections of constant length, e.g, 100 (except for the last one, which may be shorter).

**Comment by MSP-Greg** [comment]
I did an MSYS2 build from this PR.  Two issues:

1. Check failed, summary:
```
Test Summary Report
-------------------
03-test_fipsinstall.t            (Wstat: 1280 Tests: 23 Failed: 5)
  Failed tests:  19-23
  Non-zero exit status: 5
04-test_conf.t                   (Wstat: 512 Tests: 4 Failed: 2)
  Failed tests:  2, 4
  Non-zero exit status: 2
25-test_x509.t                   (Wstat: 512 Tests: 12 Failed: 2)
  Failed tests:  6-7
  Non-zero exit status: 2
Files=209, Tests=3178, 2738 wallclock secs ( 9.70 usr  1.42 sys + 432.00 cusr 877.04 csys = 1320.16 CPU)
Result: FAIL
```

2. When creating the package/install, LICENSE couldn't be found, which is now LICENSE.txt?  Copied & renamed, and install completed.

Haven't tried to use the package yet.  MSYS2 gcc is 10.2.0.

File: Configurations/unix-Makefile.tmpl [hunk 1]
```diff
@@ -1506,11 +1506,12 @@ EOF
       my %args = @_;
       my $lib = platform->staticlib($args{lib});
       my @objs = map { platform->obj($_) } @{$args{objs}};
-      my $objs = join(" \\\n" . ' ' x (length($lib) + 2),
+      my $deps = join(" \\\n" . ' ' x (length($lib) + 2),
                       fill_lines(' ', $COLUMNS - length($lib) - 2, @objs));
+      my $fill_lib = join("\n\t", (map { "\$(AR) \$(ARFLAGS) $lib $_" } @objs));
```
Comment (DDvO) [inline Configurations/unix-Makefile.tmpl:1511]: For efficiency, better split the list of object file into just a view pieces.

**Review by DDvO — CHANGES_REQUESTED** [review]

File: Configurations/unix-Makefile.tmpl (hunk 1 above)
Comment (DDvO) [reply Configurations/unix-Makefile.tmpl:1511]: @levitte, here is a solution for the efficiency problem, splitting the `ar` call to 250 objects at a time:
*  Restore the original
```
ARFLAGS         => "r",
```
* Replace in your hitherto PR:
```
      my $fill_lib = join("\n\t", (map { "\$(AR) \$(ARFLAGS) $lib $_" } @objs));
```
by
```
      my $max_per_call = 250;
      my @objs_grouped;
      push @objs_grouped, join(" ", splice @objs, 0, $max_per_call) while @objs;
      my $fill_lib = join("\n\t", (map { "\$(AR) \$(ARFLAGS) $lib $_" } @objs_grouped));
```

**Comment by DDvO** [comment]
@levitte, can we finish this up?
It's just a matter of using the solution improvement given above.

**Comment by petrovr** [comment]
About response files . The right question here is not if every ar support such functionality. 

Off topic: Lets look into one file  (look like generated for different  platform)  - https://www.cs.cmu.edu/afs/cs/project/cmcl/link.vcnectar/gnu960/src/gdbbfd/sun4/Makefile .
Lets read this "# For DOS: CFLAGS is too long for the command line; make a response file." and to see timestamp - $Id: Makefile,v 1.32 1995/12/15 22:33:34 paulr Exp $ .
The issue is command line is  so old.

Perhaps now my comment  "Functionality exist event before gcc ." is more clear.
GNU compiler and binutils are designed for unix like OS. It  is expected to support POSIX standard were if I remember well limit is128K. Linux 64-bit may be is with 4M. Exact values are not so important.

Even today Microsoft is too limited on windows 10 limit is around 8300-8400 if I remember well.
If is far from usable values for every mid-range project. More over limit is less then limit for one UNC file name (32k or 64k?).

So GNU tools windows port must implement work-around. If this was done GCC 4.5 or 4.7 .
But Richard point AR command. Let see expandargv comment in this : https://sourceware.org/git/?p=binutils.git;a=commit;f=libiberty/argv.c;h=0387ce11761e2bfb674b6e64606ea97b58e7409e 

And then use of expandargv in this :  https://sourceware.org/git/?p=binutils.git;a=commit;f=binutils/ar.c;h=f56b99f97a41876aa83f137c6526dc015980b736

Is commit timestamp enough argument to consider "response files" as valuable solution? 
It is not about just mingw*. It should work on VS build as well.

**Comment by petrovr** [comment]
About response files . The right question here is not if every ar support such functionality. 

Off topic: Lets look into one file  (look like generated for different  platform)  - https://www.cs.cmu.edu/afs/cs/project/cmcl/link.vcnectar/gnu960/src/gdbbfd/sun4/Makefile .
Lets read this "# For DOS: CFLAGS is too long for the command line; make a response file." and to see timestamp - $Id: Makefile,v 1.32 1995/12/15 22:33:34 paulr Exp $ .
The issue is command line is  so old.

Perhaps now my comment  "Functionality exist event before gcc ." is more clear.
GNU compiler and binutils are designed for unix like OS. It  is expected to support POSIX standard were if I remember well limit is128K. Linux 64-bit may be is with 4M. Exact values are not so important.

Even today Microsoft is too limited on windows 10 limit is around 8300-8400 if I remember well.
If is far from usable values for every mid-range project. More over limit is less then limit for one UNC file name (32k or 64k?).

So GNU tools windows port must implement work-around. If this was done GCC 4.5 or 4.7 .
But Richard point AR command. Let see expandargv comment in this : https://sourceware.org/git/?p=binutils.git;a=commit;f=libiberty/argv.c;h=0387ce11761e2bfb674b6e64606ea97b58e7409e 

And then use of expandargv in this :  https://sourceware.org/git/?p=binutils.git;a=commit;f=binutils/ar.c;h=f56b99f97a41876aa83f137c6526dc015980b736

Is commit timestamp enough argument to consider "response files" as valuable solution? 
It is not about just mingw*. It should work on VS build as well.

**Comment by rsbeckerca** [comment]
This PR is needed for the HPE NonStop port. There is no way to fix the environment size or maximum command line length to support the length needed to build the archive as it stands.

**Comment by rsbeckerca** [comment]
Just a thought: what about writing the object file names into a temp file and using @file: 
`-rm temp`
`echo "filename" >>temp` repeated for each file
`$(AR) $(ARFLAGS) libcrypto.a @temp`
`rm temp`

**Comment by levitte** [comment]
@petrovr, [*current* FreeBSD ar](https://www.freebsd.org/cgi/man.cgi?query=ar&sektion=1) doesn't support response files.  It has a different mechanism, MRI librarian command files.

So like I said, not every ar.  Why should we tie ourselves to GNU stuff everywhere?

**Comment by levitte** [comment]
> Just a thought: what about writing the object file names into a temp file and using @file:

Not universally supported, like has already been discussed.

**Comment by rsbeckerca** [comment]
> 
> 
> > Just a thought: what about writing the object file names into a temp file and using @file:
> 
> Not universally supported, like has already been discussed.

My apologies on that one. I missed that point.

File: Configurations/unix-Makefile.tmpl (hunk 1 above)
Comment (levitte) [reply Configurations/unix-Makefile.tmpl:1511]: Other than that, sure

Comment (levitte) [reply Configurations/unix-Makefile.tmpl]: Done

**Comment by levitte** [comment]
> My apologies on that one. I missed that point.

No worries

**Review by DDvO — APPROVED** [review]
LGTM and works as expected.

File: Configurations/unix-Makefile.tmpl (hunk 1 above)
Comment (DDvO) [reply Configurations/unix-Makefile.tmpl]: > Not sure I understand why "r" is preferable to "q" ("c" can be appended either way)

I don't know either - I just thought that in doubt the original "r" may be kept.
At least in our case both variants work.

**Comment by petrovr** [comment]
Hi Richard,

Richard Levitte wrote:
> @petrovr, [*current* FreeBSD ar](https://www.freebsd.org/cgi/man.cgi?query=ar&sektion=1) doesn't support response files.  It has a different mechanism, MRI librarian command files.
> 
> So like I said, not every ar.  Why should we tie ourselves to GNU stuff everywhere?
> 

GNU stuff!?

I'm not sure why we start to talk for BSD as this issue pretend to fix "12116 MinGW build fails due to overlong ar argument list".

On MS-DOS, IBM-DOS, Microsoft Windows "response file" is technology used long time ago - before GNU Linux, before FreeBSD.
Microsoft, Intel (Borland,  Watcom) ...
For instance
  https://docs.microsoft.com/en-us/cpp/build/reference/linker-options?view=vs-2019
  https://scc.ustc.edu.cn/zlsc/sugon/intel/compiler_c/main_cls/bldaps_cls/common/bldaps_using_options.htm

In 2005 finally GNU binutils adds support for "@response file".
It was late - 6 or 7 years after first MinGW release but required functionally.

The fact that "@response file" is not only for MinGW is out of scope - this PR is for MinGW.


But if is removed "Fixes #12116" then why to change build?
If is fine for HPE NonSTOP go ahead.
I have no idea if the tools used on HPE NonSTOP support response file.


It seems to me builds for "Microsoft Windows" OS require another PR.
Build tools for "Microsoft Windows" host support "response file" and OpenSSL build system could benefit from this.


Regards,
Roumen Petrov

**Comment by levitte** [comment]
@petrovr, this change affects unix-Makefile.tmpl, which is not limited to mingw or other POSIX emulations on windows. Therefore, I'm trying to avoid special casing for those, and do something that works uniformly in *all* Unix like environments.

So, the question is, does this PR solve the issue? That should be sufficient.

**Comment by rsbeckerca** [comment]
The final build of libcrypto.so still uses the entire list of .o files instead of libcrypto.a. So I'm not sure this fix is complete.

**Comment by rsbeckerca** [comment]
> 
> 
> The final build of libcrypto.so still uses the entire list of .o files instead of libcrypto.a. So I'm not sure this fix is complete.

I had to hack the makefile to use the -Weld_obey=file option on this platform (not a good generic option), and run the eld (the equivalent of ld) directly instead of using CC. This is not a sustainable approach, but I did manage to get libcrypto.so built. If using CC, one .o file has to be specified on the command line or c99 does not know what to do with its argument list. What I ended up with was:

```
        <repeat the echo for all files going into libcrypto.so>
        echo providers/libcrypto-shlib-baseprov.o >> libcrypto.in
        echo providers/libcrypto-shlib-defltprov.o >> libcrypto.in
        echo providers/libcrypto-shlib-nullprov.o >> libcrypto.in
        $(CC) $(LIB_CFLAGS) $(LIB_LDFLAGS) \
                -Wshared \
                -o libcrypto.so \
                -Weld_obey=libcrypto.in \
                providers/libimplementations.a providers/libcommon.a providers/libnonfips.a  $(LIB_EX_LIBS)
```

**Comment by DDvO** [comment]
> The final build of libcrypto.so still uses the entire list of .o files instead of libcrypto.a. So I'm not sure this fix is complete.

The problem was only with the `ar` tool, which is used for producing `libcrypto.a`, not for `libcrypto.so`.
I tried out the fix on my MinGW installation (which has the restricted version of `ar`), and it turned out be be sufficient there.

**Comment by rsbeckerca** [comment]
> 
> 
> > The final build of libcrypto.so still uses the entire list of .o files instead of libcrypto.a. So I'm not sure this fix is complete.
> 
> The problem was only with the `ar` tool, which is used for producing `libcrypto.a`, not for `libcrypto.so`.
> I tried out the fix on my MinGW installation (which has the restricted version of `ar`), and it turned out be be sufficient there.

Sadly, the environment limits that case the `ar` issue on NonStop are the same for `c99` or any other program.

**Comment by rsbeckerca** [comment]
@DDvO Do we need a new issue for the libcrypto.so build limit? I can create one.

**Comment by DDvO** [comment]
> @DDvO Do we need a new issue for the libcrypto.so build limit? I can create one.

Better create a new issue to keep matters separate.
The fix done here is already approved and is already sufficient for fixing the original issue.

**Comment by levitte** [comment]
I'm amazed that the libcrypto.so command line would be shorter than the libcrypto.a command line.  It doesn't make sense.  But maybe I've missed something.

Anyway, I wouldn't be surprised if we'd hit a limit there as well with mingw

**Comment by rsbeckerca** [comment]
> 
> 
> I'm amazed that the libcrypto.so command line would be shorter than the libcrypto.a command line. It doesn't make sense. But maybe I've missed something.
> 
> Anyway, I wouldn't be surprised if we'd hit a limit there as well with mingw

I'm building using #12706 so that helps with libcrypto.a.

**Comment by DDvO** [comment]
> I'm amazed that the libcrypto.so command line would be shorter than the libcrypto.a command line. It doesn't make sense. But maybe I've missed something.

The point is that, as I commented earlier today, 
`libcrypto.a` is constructed using `ar`, while `libcrypto.so` does not use `ar` (but `gcc`).

**Comment by rsbeckerca** [comment]
> 
> 
> > I'm amazed that the libcrypto.so command line would be shorter than the libcrypto.a command line. It doesn't make sense. But maybe I've missed something.
> 
> The point is that, as I commented earlier today,
> `libcrypto.a` is constructed using `ar`, while `libcrypto.so` does not use `ar` (but `gcc`).

I understand that, part. Our compiler is c99, not gcc (not really relevant). The limitation is the platform, not the compiler. Before applying #12706, the command lines were comparable +/- 1Kb with options. Now they are not, so the libcrypto.so command line is hitting the same limit as ar did previously. I opened #12667 for the libcrypto.so situation.

**Comment by makr** [comment]
Just to mention it: _ar_ on AIX doesn't support response files. Response files for _ar_ are also not mentioned in the POSIX standard.

**Comment by levitte** [comment]
> Response files for _ar_ are also not mentioned in the POSIX standard.

Thank you for making that point

**Comment by paulidale** [comment]
Why not use `xargs(1)` for this?

**Comment by rsbeckerca** [comment]
> 
> 
> Why not use `xargs(1)` for this?

While `xargs(1)` has an effectively unlimited input size, it is still restricted to the platform's limit on how long a command can be. This is the issue with `ar` and `c99`, which is really subsumed by a fundamental problem of size limits on forming a shell that is smaller than what is generated from the contents of the Makefile. Good thought though.

**Comment by paulidale** [comment]
The whole point is that xargs knows the limit and won't exceed it: `cat files | xargs ar q output.a`

**Comment by rsbeckerca** [comment]
> 
> 
> The whole point is that xargs knows the limit and won't exceed it: `cat files | xargs ar q output.a`

I don't see how this will help either `ar` or `c99`, but that might by my own lack of understanding of xargs or the implementation on my platform. You'll just get a truncated library/DLL. While you can add to an archive, this is not a general solution for an arbitrary product. Although perhaps `xargs` would run `ar` multiple times or some platforms but not NonStop. Regardless, the `ar` situation is resolved effectively by the fix described in this pull request (levitte-12116) even if it slows down the build somewhat.

**Comment by paulidale** [comment]
Yes, `xargs` runs multiple times ensuring that the command line length limits are respected.

**Comment by slontis** [comment]
@levitte - is this one mergeable - I have updated the tag to approved.

**Comment by levitte** [comment]
@DDvO approved two days ago, and it hasn't changed since, so it's mergable.

(@DDvO, has no one told you about the labeling we do?  We have a bot that picks up on those)

**Comment by levitte** [comment]
Merged

master:
385deae79f26dd685339d3141a06d04d6bd753cd Building: Build Unix static libraries one object file at a time

I've skipped 1.1.1, the cherry-pick wasn't clean enough, so that requires a separate backport PR.  Anyone is welcome to produce one.

**Comment by DDvO** [comment]
> (@DDvO, has no one told you about the labeling we do? We have a bot that picks up on those)

I know about the merge-related labels, but I'm simply not used to setting them (because I'm typically not the one who is able to give the final approval) therefore I forgot to set the `approval:done` and `approval:ready to merge` flags.

**Comment by slontis** [comment]
The 'master' and 'needs approval' labels would also be a good habit of adding to your PR's. Just adding those to all your existing PR's took a while :)

**Comment by DDvO** [comment]
> > > I'm amazed that the libcrypto.so command line would be shorter than the libcrypto.a command line. It doesn't make sense. But maybe I've missed something.
> > 
> > 
> > The point is that, as I commented earlier today,
> > `libcrypto.a` is constructed using `ar`, while `libcrypto.so` does not use `ar` (but `gcc`).
> 
> I understand that, part. Our compiler is c99, not gcc (not really relevant). The limitation is the platform, not the compiler. Before applying #12706, the command lines were comparable +/- 1Kb with options. Now they are not, so the libcrypto.so command line is hitting the same limit as ar did previously. I opened #12667 for the libcrypto.so situation.

@rsbeckerca, #12667 is on a different issue.
And the fix done here does solve all the build issues with MinGW, so fixes #12116.
BTW libcrypto has the suffix `-3-x64.dll` there.

**Comment by DDvO** [comment]
> Regardless, the `ar` situation is resolved effectively by the fix described in this pull request (levitte-12116) even if it slows down the build somewhat.

Yes. I meanwhile found that we could have put the split after 500 rather than 250 objects,
which would have made the solution somewhat more efficient, but the difference should be minimal.

**Comment by DDvO** [comment]
@levitte, I've just submitted a minor fixup PR #12821.

**Pull request closed without merging**